pvlib/iotools/era5.py

```diff
@@ -0,0 +1,277 @@
+    *Mapped field names are returned when the map_variables argument is True*
+    ---------------------------------------------------------------------------
+    2tm, temp_air             float   Air temperature at 2 m above ground [K]
+    u10                       float   Horizontal airspeed towards east at 10 m [m/s]
+    v10                       float   Horizontal airspeed towards north at 10 m [m/s]
+    sp, pressure              float   Atmospheric pressure at the ground [Pa]
+    msdwswrf, ghi             float   Mean surface downward short-wave radiation flux [W/m^2]
+    msdwswrfcs, ghi_clear     float   Mean surface downward short-wave radiation flux, clear sky [W/m^2]
+    msdrswrf, bhi             float   Mean surface direct short-wave radiation flux [W/m^2]
+    msdrswrfcs, bhi_clear     float   Mean surface direct short-wave radiation flux, clear sky [W/m^2]
+    ========================  ======  =========================================
+
+    Returns
+    -------
+    data: DataFrame
+        ERA5 time-series data, fields depend on the requested data. The
+        returned object is either a pandas DataFrame or an xarray dataset,
+        depending on the output_format parameter.
+    metadata: dict
+        Metadata for the time-series.
+
+    See Also
+    --------
+    pvlib.iotools.read_era5
+
+    References
+    ----------
+    .. [1] `ERA5 hourly data on single levels from 1979 to present
+       <https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels?tab=overview>`_
+    .. [2] `ERA5 data documentation
+       <https://confluence.ecmwf.int/display/CKB/ERA5%3A+data+documentation>`_
+    .. [3] `How to use the CDS API
+       <https://cds.climate.copernicus.eu/api-how-to>`_
+    .. [4] `CDSAPI source code
+       <https://github.com/ecmwf/cdsapi>`_
+    .. [5] `Climate Data Store (CDS) API Keywords
+       <https://confluence.ecmwf.int/display/CKB/Climate+Data+Store+%28CDS%29+API+Keywords>`_
+    .. [6] `Climate Data Storage user registration
+       <https://cds.climate.copernicus.eu/user/register>`_
+    """  # noqa: E501
+    if cdsapi is None:
+        raise ImportError('Retrieving ERA5 data requires cdsapi to be installed.')  # noqa: E501
+
+    cds_client = cdsapi.Client(url=CDSAPI_URL, key=api_key,
+                               verify=1, quiet=True)
+
+    # Area is selected by a box made by the four coordinates: [N, W, S, E]
+    try:
+        area = [latitude[1], longitude[0], latitude[0], longitude[1]]
+    except TypeError:
+        area = [latitude+0.005, longitude-0.005,
+                latitude-0.005, longitude+0.005]
+
+    params = {
+        'product_type': product_type,
+        'variable': variables,
+        'date': start.strftime('%Y-%m-%d') + '/' + end.strftime('%Y-%m-%d'),
+        'time': ERA5_HOURS,
+        'grid': grid,
+        'area': area,
+        'format': 'netcdf'}
+
+    # Retrieve remote path to the file
```

Comment:
Yes, it is better to avoid having the user download the same thing more than once.

There is a potential danger in this function: if an exception is raised while saving or reading/conversion, the in-memory data will be lost.